compiler-core/src/language_server/engine.rs

```diff
@@ -205,6 +205,43 @@ where
             let byte_index =
                 line_numbers.byte_index(params.position.line, params.position.character);
 
+            // Use binary search to find if the byte_index is
+            // in the range [span.start, span.end]
+            //
+            // Do note that the end is inclusive, because we would like to
+            // detect if the cursor is right after the comment
+            // (when we are typing/insert mode in Vim)
+            let comment_cmp = |span: &SrcSpan| {
+                if byte_index < span.start {
+                    Ordering::Less
+                } else if span.end < byte_index {
+                    // inclusive end
+                    Ordering::Greater
+                } else {
+                    Ordering::Equal
+                }
+            };
+
+            // If in comment context, do not provide completions
+            //
+            // We cannot simply use `module.extra.is_within_comment` as it
+            // cannot detect if we are currently typing/inserting text at the
+            // end of the comment
+            if module.extra.comments.binary_search_by(comment_cmp).is_ok()
+                || module
+                    .extra
+                    .doc_comments
+                    .binary_search_by(comment_cmp)
+                    .is_ok()
+                || module
+                    .extra
+                    .module_comments
+                    .binary_search_by(comment_cmp)
+                    .is_ok()
+            {
+                return Ok(None);
+            }
+
```

Comment:
@lpil Moving the discussion here after I made some changes to cover all edge cases.

Reply:
Instead of "is this cursor in a comment?", we should view this as "if I insert a char in this cursor, would it be included in the comment?". If true, then do not provide completions.

To demonstrate the new POV, suppose that we have a comment with 2 characters in it

```
//ab
  ^^^ don't provide completions if our cursor is in any of these positions
```

Expanded version:

```
//|ab
  ^ cursor is at byte index 2 (no completion)
//a|b
   ^ cursor is at byte index 3 (no completion)
//ab|
    ^ cursor is at byte index 4 (no completion)
```

So, with the new POV, we should check if the byte index is in the range `[start, end]` (note the inclusive end)

Reply:
To compare:

```
is_within_comment:

//abc
  ^^^ considers this span
```

```
new logic to avoid completions in comments:

//abc
  ^^^^ considers this span (extended by 1)
```

Reply:
If the `is_within_comment` has the same logic as the new one, then this wouldn't make much sense:

Suppose that the source code is only `//abc`:

```
//abc
     ^ is_within_comment(5) -> true (?)
```

But the source code only has 5 characters, so if we index the source code with 5 (e.g. `code[5]`), it would panic

Reply:
Once we have a common understanding, I'll put the updated explanation in the implementation

Reply:
You've not included an example for when the cursor should not be considered in a comment? All of the examples you've given are within a comment.

Reply:
Technically, the definition of "is_within_comment(byte_index)" is `comment_span.start <= byte_index < comment_span.end`

So the cursor in `//abc|` (where the character `|` refers to the cursor) is not in a comment (if we use `is_within_comment`)

Reply:
And what is the code example that makes that true? All the examples you've given are evidence that it is not the case, indicating an off-by-one error.

Reply:
Examples of correct cases (`is_within_comment` returns true):

```
//|abc
//a|bc
//ab|c
```

Reply:
Right, and what's the counterexample that shows that the modifications you've made outside of the function shouldn't be moved inside the function?

Reply:
The modifications made outside of the function are in green (added, which is a special case used for ignoring the completion), and the line in red is the one from `SrcSpan::cmp_byte_index`

```diff
  if byte_index < span.start {
      Ordering::Less
- } else if span.end <= byte_index {
+ } else if span.end < byte_index {
+    // inclusive end
      Ordering::Greater
  } else {
      Ordering::Equal
  }
```

Reply:
Again, please give the examples that explain why a special case is needed. Thank you.

Reply:
So, a special case is needed here to cover a case whenever we ask for completions right after `//`.

Using only `is_within_comment` here triggers completions right after `//`:

https://github.com/gleam-lang/gleam/assets/65026286/63404313-c72f-41c7-a2df-a7ea129b688a


But using the special case doesn't trigger completions right after `//`:

https://github.com/gleam-lang/gleam/assets/65026286/481c3a49-f083-48e5-a7c5-89a1221a2611

Reply:
I'm not asking for examples of when this is needed for comments, I'm asking for how it is not needed for the other uses of this function.

Reply:
I don't think there are any other uses for this "modified" version for now.

Unless, in the future, we need to detect whether we are in a comment during insert mode (vim) for some LSP-related features.

Reply:
Bumping this in case you missed this @lpil

Reply:
The existing uses are only for detecting if a char is inside a comment context for deleting redundant tuples in case exprs (https://github.com/gleam-lang/gleam/blob/main/compiler-core/src/language_server/code_action.rs#L242).

The new behaviour doesn't suit the existing uses because of this one specific case:

In raw/escaped chars,

```gleam
// comment\n0
          ^^this is a newline
```

which is basically,

```gleam
// comment
0
```

The comment span only covers `// comment`. But if we use the new behaviour, it would say that the newline (`\n`) is within comment, i.e. `is_within_comment() -> true` (basically took more characters than it should).

Reply:
Could you give an example of when the new behaviour is not suitable for the other uses please.